feat: add new metrics

edoardomich opened the pull request:
Enhance the coverage by including additional metrics introduced in version 7.0.x series, as well as new ones added in the recently released version 8.0.0. Additionally, I propose adding a flag to retrieve global metrics that represent the aggregated totals of the per-thread counters.

The tests were conducted using simple container images with Suricata v6.0.20, v7.0.11, and the newly released v8.0.0

edoardomich committed:
feat: add additional metrics

Add additional metrics for version 7.0.x and the upcoming 8.0.0 and correct small typos and compatability problems with the old 6.0.20

edoardomich committed:
feat: add on demand Total metrics

Add also the option to expose overall total metrics defined at the top-level as the sum of all the per-thread ones, avoiding more computation on the receiving side

edoardomich committed:
test: correct TestDump604Netmap

aligned with the total number of new metrics

awelzel commented on main.go:
```diff
@@ -83,7 +83,7 @@ var (
 		// The following 4 are put into a single metrics afpacket_polls_total
 		// where the result is a labels.
-		// newPerThreadCounterMetric("capture", "afpacket_poll_total", "", "polls"),
+		newPerThreadCounterMetric("capture", "afpacket_poll_total", "", "polls"),
```
Mind reverting this one? The idea here was that summing the `afpacket_poll_results_total` counters should result in the total. If this is not the case anymore, is `perThreadAFPacketPollResultEntries` missing an entry for a different result?

This is from the current suricata dev version and it  seems signal, timeout, data, error capture polls.
```
    "capture": {
      "kernel_packets": 18574,
      "kernel_drops": 0,
      "errors": 0,
      "afpacket": {
        "busy_loop_avg": 0,
        "polls": 18422,
        "poll_signal": 0,
        "poll_timeout": 16464,
        "poll_data": 1958,
        "poll_errors": 0,
        "send_errors": 0
      }
```

awelzel commented on main.go:
```diff
@@ -150,17 +150,35 @@ var (
+	// From: .thread.decoder.event.afpacket
+	// New in 8.0.0: d78f2c9a4e2b59f44daeddff098915084493d08d
+	perThreadDecoderEventAFPacketMetrics = []metricInfo{
+		newPerThreadCounterMetric("decoder_event", "afpacket_trunc_pkt", "", "trunc_pkt").Optional(),
```
```suggestion
		newPerThreadCounterMetric("decoder_event", "afpacket_truncated_packets_total", "", "trunc_pkt"),
```
Mostly to align with the other decoder packet counter metrics above (udp_packets_total) and dropping `Optional()` as it seems if  `afpacket` is within `decoder.event`, can expect `trunc_pkt` to be included.

---

Aside, seems a bit unfortunate the metric is not part of the `capture.afpacket` object as the event is raised in `source-af-packet.c`, but not sure it's worth an upstream ticket to move or also include it there.

@jlucovsky / @jasonish, thoughts? Nothing urgent or to do, just some feedback / food for thought, for the `afpacket.trunc_pkt` stat.

edoardomich replied:
Ok for the rename. I also removed the Optional()

awelzel commented on main.go:
```diff
@@ -150,17 +150,35 @@ var (
 	perThreadFlowMetrics = []metricInfo{
+		// New in 7.0.0: b0993d6fd8ffb6abafdea9324cd0166106af1851
+		newPerThreadCounterMetric("flow", "flow_total", "", "total").Optional(),
+		newPerThreadCounterMetric("flow", "active_total", "", "active").Optional(),
```
This is a gauge.
```suggestion
		newPerThreadGaugeMetric("flow", "active_flows", "", "active").Optional(),
```

awelzel commented on main.go:
```diff
@@ -150,17 +150,35 @@ var (
 	// From .thread.flow
 	perThreadFlowMetrics = []metricInfo{
+		// New in 7.0.0: b0993d6fd8ffb6abafdea9324cd0166106af1851
+		newPerThreadCounterMetric("flow", "flow_total", "", "total").Optional(),
```
```suggestion
		newPerThreadCounterMetric("flow", "all_total", "", "total").Optional(),
```

Just to prevent `flow_flow_total` and :crossed_fingers: for there never being an "all" metric :-)

awelzel commented on main.go:
```diff
@@ -173,7 +191,37 @@ var (
 		newPerThreadCounterMetric("flow_wrk", "flows_evicted_total", "", "flows_evicted"),
 		newPerThreadCounterMetric("flow_wrk", "flows_injected_total", "", "flows_injected"),
-		newPerThreadCounterMetric("flow_wrk", "flows_evicted_needs_work", "", "flows_evicted_needs_work"),
+		newPerThreadCounterMetric("flow_wrk", "flows_injected_max_total", "", "flows_injected_max").Optional(),
```
```suggestion
		newPerThreadGaugeMetric("flow_wrk", "flows_injected_max", "", "flows_injected_max").Optional(),
```

Other `max` metrics are gauges which is a bit better, because it doesn't make sense to compute a `rate()` on these. Making these counters and using the `_total` suffix can be misleading.

awelzel commented on main.go:
```diff
@@ -173,7 +191,37 @@ var (
 		newPerThreadCounterMetric("flow_wrk", "flows_evicted_pkt_inject_total", "", "flows_evicted_pkt_inject"),
 		newPerThreadCounterMetric("flow_wrk", "flows_evicted_total", "", "flows_evicted"),
 		newPerThreadCounterMetric("flow_wrk", "flows_injected_total", "", "flows_injected"),
-		newPerThreadCounterMetric("flow_wrk", "flows_evicted_needs_work", "", "flows_evicted_needs_work"),
```
Oh, that's a duplicate? Thanks!

awelzel commented on main.go:
```diff
@@ -660,6 +744,34 @@ func handleTransmitThread(ch chan<- prometheus.Metric, threadName string, thread
+		}
+	}
+
+	if state, ok := flow["state"].(map[string]any); ok {
```
```suggestion
	if state, ok := end["state"].(map[string]any); ok {
```

This needs to be inside the above `if`, else `end` is not defined. Or switch it to an early return style, either is fine for me.

awelzel commented on main.go:
```diff
@@ -660,6 +744,34 @@ func handleTransmitThread(ch chan<- prometheus.Metric, threadName string, thread
+		}
+	}
+
+	if tcp_state, ok := flow["tcp_state"].(map[string]any); ok {
```
```suggestion
	if tcp_state, ok := end["tcp_state"].(map[string]any); ok {
```

awelzel commented on main.go:
```diff
@@ -769,6 +896,16 @@ func handleGlobal(ch chan<- prometheus.Metric, message map[string]any) {
+			}
+		}
+	} else {
+		log.Printf("WARN: No top-level flow entry message")
```
```suggestion
		log.Printf("WARN: No top-level defrag entry message")
```

awelzel commented on main.go:
```diff
@@ -813,6 +950,14 @@ func handleGlobal(ch chan<- prometheus.Metric, message map[string]any) {
 		log.Printf("WARN: No top-level detect entry")
 	}
 
+	if *totals {
```
Yeah, that seems fine. I wonder if *totals should disable per thread metrics? I could see someone just wanting totals instead of per thread detail to have less data, but now it's even more :-)

Do you think you could rework it such that perThread metrics are ignored if *totals is set?

awelzel commented on main.go:
```diff
@@ -255,6 +314,12 @@ var (
 
+	// From .thread.file_store
+	perThreadFileStoreMetrics = []metricInfo{
+		newPerThreadCounterMetric("filestore", "open_files_max_hit", "", "open_files_max_hit").Optional(),
```
Running a default configuration, I only see this in dump-counters:

```
        "file_store": {
          "open_files": 0
        },
```

Mind posting a section of your dump-counters that includes the filestore metrics?

The max_hit should probably also be a guage.

edoardomich replied:
You are right, it's a Gauge. It's optional because it appears in the dump of the counters only if file store is enabled in the configuration:
```
  - files-store:
      version: 2
      enabled: yes
      [...]
```

`open_files_max_hit`, is a per-thread metric under `file_store`, and can be seen in the [schema.json](https://github.com/OISF/suricata/blob/2e69e0d5c16c35ea05497c378fa03908824fe5e4/etc/schema.json#L6566)

<img width="359" height="221" alt="image" src="https://github.com/user-attachments/assets/4f42e6c0-7d01-44a5-9d71-8557a42eb3d5" />

There is also a _total_ one under `message.file_store`

awelzel reviewed (changes requested):
Hey @edoardomich , thanks for your contribution. Mind checking out the comments and see if you agree with them and adapt the changes?

If you could add a suricata dump-counters output into testdata for 8.0.x with the configuration you used to develop this, that'd be awesome! Maybe also add a smoke test for the  `flow_end` merics that seemed to be non-functional. The tests aren't the best, but hopefully sufficient for adding another smoke test.

Thanks!

edoardomich committed:
test: add new test for version 8.0.0

add `TestDump800AFPacket` and `TestDump800AFPacketFileStore` to test new metrics

edoardomich commented:
I corrected the typos and the suggested changes. I also add some tests for the new metrics.
Let me know  if I had to perform new changes or I can squash the commits for the merge

awelzel commented on README.md:
```diff
@@ -17,6 +17,8 @@ Usage of ./suricata_exporter:
   -quiet
         supress logging messages when suricata is not enabled
+  -totals
+        Export the overall global total metrics. (default false)
```
```suggestion
        Export only the overall global total metrics instead of per-thread metrics. (default false)
```

awelzel commented on main.go:
```diff
@@ -255,6 +314,12 @@ var (
+	// From .thread.file_store
+	perThreadFileStoreMetrics = []metricInfo{
+		newPerThreadGaugeMetric("filestore", "open_files_max_hit", "", "open_files_max_hit").Optional(),
+		newPerThreadCounterMetric("filestore", "fs_errors", "", "fs_errors").Optional(),
```
```suggestion
		newPerThreadCounterMetric("filestore", "fs_errors_total", "", "fs_errors").Optional(),
```

awelzel reviewed (commented):
This looks great. There's two more minor nits. Please squash the fixups together and I'll merge it into master and cut a release.

Thanks a lot!

awelzel commented:
@edoardomich - thank you for your contribution! It's released with `v0.5.0` now :+1:
